sdk/storage/Azure.Storage.Common/src/Shared/PartitionedDownloader.cs

```diff
@@ -0,0 +1,406 @@
+        /// The maximum number of simultaneous workers.
+        /// </summary>
+        private readonly int _maxWorkerCount;
+
+        /// <summary>
+        /// The size of the first range requested (which can be larger than the
+        /// other ranges).
+        /// </summary>
+        private readonly long _initialRangeSize;
+
+        /// <summary>
+        /// The size of subsequent ranges.
+        /// </summary>
+        private readonly long _rangeSize;
+
+        /// <summary>
+        /// The name of the calling operaiton.
+        /// </summary>
+        private readonly string _operationName;
+
+        public PartitionedDownloader(
+            Behaviors behaviors,
+            StorageTransferOptions transferOptions = default,
+            string operationName = null)
+        {
+            // TODO: Check on differences between defaults for blobs/files
+
+            // Modifying conditions and manually checking the Etag for changes are behaviors
+            // unique to particular services and can use a no-op; the rest are required
+            _modifyConditions = behaviors.ModifyConditions ?? ModifyConditionsNoOp;
+            _checkForEtagChange = behaviors.CheckForEtagChange ?? CheckEtagNoOp;
+            _singleDownloadInternal = behaviors.SingleDownload
+                ?? throw Errors.ArgumentNull(nameof(behaviors.SingleDownload));
+            _copyToAsync = behaviors.CopyToAsync
+                ?? throw Errors.ArgumentNull(nameof(behaviors.CopyToAsync));
+            _copyTo = behaviors.CopyTo
+                ?? throw Errors.ArgumentNull(nameof(behaviors.CopyTo));
+            _createScope = behaviors.Scope
+                ?? throw Errors.ArgumentNull(nameof(behaviors.Scope));
+
+            // Set _maxWorkerCount
+            if (transferOptions.MaximumConcurrency.HasValue
+                && transferOptions.MaximumConcurrency > 0)
+            {
+                _maxWorkerCount = transferOptions.MaximumConcurrency.Value;
+            }
+            else
+            {
+                _maxWorkerCount = Constants.Blob.Block.DefaultConcurrentTransfersCount;
+            }
+
+            // Set _rangeSize
+            if (transferOptions.MaximumTransferSize.HasValue
+                && transferOptions.MaximumTransferSize.Value > 0)
+            {
+                _rangeSize = Math.Min(transferOptions.MaximumTransferSize.Value, Constants.Blob.Block.MaxDownloadBytes);
+            }
+            else
+            {
+                _rangeSize = Constants.DefaultBufferSize;
+            }
+
+            // Set _initialRangeSize
+            if (transferOptions.InitialTransferSize.HasValue
+                && transferOptions.InitialTransferSize.Value > 0)
+            {
+                _initialRangeSize = transferOptions.InitialTransferSize.Value;
+            }
+            else
+            {
+                _initialRangeSize = Constants.Blob.Block.DefaultInitalDownloadRangeSize;
+            }
```

Comment:
I'd extract the logic that populates defaults to blobs and shares and accept a struct/class that has them all figured out in the downloader. I'm not sure if defaults for blobs should be same for shares.